From 03735122bf7d46c2612b98de828c2931a25c039e Mon Sep 17 00:00:00 2001
From: Daniel Baluta <daniel.baluta@nxp.com>
Date: Wed, 23 May 2018 09:40:55 +0300
Subject: [PATCH 4374/5242] MLK-17481-3: ASoC: fsl: Fix DSP memory mappings

commit  584c308b99256fcb6d1a45eab9da679c817740f5 from
https://source.codeaurora.org/external/imx/linux-imx.git

We load DSP firmware from the ARM side at 0x556e8000 but because the
compiler generated memory layout starts at 0x596e8000 we need to do
some fixups.

Thus, each address (in DSP local memory) generated by the compiler
needs to be substracted an offset = 0x596e8000 - 0x556e8000 = 0x4000000.

Because this only happens on QM we will use dts to specify the offset.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>
Reviewed-by: Shengjiu Wang <shengjiu.wang@nxp.com>
(cherry picked from commit 8d4518d2a5d956549e829470af15003d7adff841)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/freescale/fsl-imx8qm-device.dtsi      |    1 +
 sound/soc/fsl/fsl_dsp.c                            |    8 ++++++++
 sound/soc/fsl/fsl_dsp.h                            |    2 ++
 3 files changed, 11 insertions(+)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi
index b1066c8..ea5a8a3 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi
@@ -3644,6 +3644,7 @@
 			<&clk IMX8QM_AUD_DSP_CORE_CLK>;
 		clock-names = "ipg", "ocram", "core";
 		fsl,dsp-firmware = "imx/dsp/hifi4.bin";
+		fixup-offset = <0x4000000>;
 		power-domains = <&pd_dsp>;
 	};
 
diff --git a/sound/soc/fsl/fsl_dsp.c b/sound/soc/fsl/fsl_dsp.c
index 3bdf600..0a1f2a4 100644
--- a/sound/soc/fsl/fsl_dsp.c
+++ b/sound/soc/fsl/fsl_dsp.c
@@ -670,6 +670,12 @@ static void dsp_load_firmware(const struct firmware *fw, void *context)
 				  (const void *)image,
 				  shdr->sh_size);
 			} else {
+				/* sh_addr is from DSP view, we need to
+				 * fixup addr because we load the firmware from
+				 * the ARM core side
+				 */
+				sh_addr -= dsp_priv->fixup_offset;
+
 				memcpy_dsp((void *)(dsp_priv->regs +
 						(sh_addr - dsp_priv->paddr)),
 						(const void *)image,
@@ -820,6 +826,8 @@ static int fsl_dsp_probe(struct platform_device *pdev)
 	ret = of_property_read_string(np, "fsl,dsp-firmware", &fw_name);
 	dsp_priv->fw_name = fw_name;
 
+	ret = of_property_read_u32(np, "fixup-offset", &dsp_priv->fixup_offset);
+
 	platform_set_drvdata(pdev, dsp_priv);
 	pm_runtime_enable(&pdev->dev);
 
diff --git a/sound/soc/fsl/fsl_dsp.h b/sound/soc/fsl/fsl_dsp.h
index 4541777..9288ac1 100644
--- a/sound/soc/fsl/fsl_dsp.h
+++ b/sound/soc/fsl/fsl_dsp.h
@@ -76,6 +76,8 @@ struct fsl_dsp {
 	dma_addr_t			 dsp_config_phys;
 	int				 dsp_config_size;
 
+	unsigned int			fixup_offset;
+
 	/* ...proxy data structures */
 	struct xf_proxy proxy;
 
-- 
1.7.9.5

